veps/sig-compute/vmtemplate-proposal.md

```diff
@@ -0,0 +1,752 @@
+            F("Imported VM disk(s)")
+            G{"VM Template"}
+        end
+        F -- "Referenced by" --> G
+        G -- "Customize & Create VMs" --> H("VM Deployment Process")
+
+        subgraph VMNamespaceD [VM Namespace D]
+            I("Deployed VirtualMachine 3")
+        end
+        subgraph VMNamespaceC [VM Namespace C]
+            J("Deployed VirtualMachine 2")
+        end
+        subgraph VMNamespaceB [VM Namespace B]
+            K("Deployed VirtualMachine 1")
+        end
+
+        H --> I
+        H --> J
+        H --> K
+    end
+```
+
+## Goals
+
+* Provide native support for in-cluster templating of golden image-based
+  VirtualMachines
+* Provide native support for in-cluster templating of existing VirtualMachines
+* Provide inter-namespace sharing of templated VirtualMachines
+* Provide inter-cluster sharing of templated VirtualMachines
+* Provide import/export of VirtualMachine templates in a standardized format
+  (OVA/OVF)
+
+## Non Goals
+
+* Replace external configuration management tools like Helm or Kustomize.
+* Real-time guest operations or application-level quiescing during the template
+  creation process are not in scope for the initial implementation
+* Adding a plugin mechanism to `virtctl` (such a mechanism could be
+  used to break dependencies in the future)
+
+## Definition of users
+
+* Cluster admin: Manages the cluster and should be able to provide templates
+  that can be consumed by VM owners
+* VM owners: Users on the cluster which should be able to create VMs in freeform
+  or from templates. They should also be able to create templates from their
+  existing VMs or import templates from a standardized format
+
+## User Stories
+
+* As a VM owner, I want the ability to create a templated VirtualMachines
+  referencing a golden image
+* As a VM owner, I want the ability to create a templated VirtualMachines from
+  an existing VirtualMachine
+* As a VM owner, I want the ability to create a VirtualMachine from a templated
+  VirtualMachine that is stored in a different namespace
+* As a VM template owner, I want the ability to import templates from a
+  standardized format (OVA/OVF)
+* As a VM template owner, I want the ability to export templates to a
+  standardized format (OVA/OVF)
+* As a cluster admin, I want to restrict users via RBAC to only create VMs
+  from approved templates
+
+## Repos
+
+* [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+* [kubevirt/common-templates](https://github.com/kubevirt/common-templates)
+* [kubevirt/common-instancetypes](https://github.com/kubevirt/common-instancetypes)
+
+## Design
+
+### Development and deployment of this feature
+
+This design suggests developing this feature outside the core
+`kubevirt/kubevirt` repository, allowing to better organize the new code and to
+keep it separate from the core KubeVirt implementation. This means the feature
+will have its own operator, API server and controller.
```

Comment:
We could potentially reuse the template for creating this from AAQ (https://github.com/kubevirt/application-aware-quota) @Barakmor1 made this into something reusable that can be extended to spawn new operators.
@Barakmor1 maybe you could chime in here :)

Reply:
Yes, the AAQ repo can definitely serve as a good starting point for a new operator. I’d be happy to help with that effort.

To be honest, I haven’t explored the details of instancetype, preference, and templates deeply. But from my perspective, supporting both feels a bit redundant. It’s unclear why we need two partial configuration CRDs and another object (template) to combine everything.

It might be clearer to have a new object, like VirtualMachinePuzzle, managed by a separate operator, that includes both and creates the final VirtualMachine.

That said, I’m not very familiar with this logic, so I might be missing something.